### PR TITLE
Vomit on users from /mod

### PR DESCRIPTION
**Pull request by jacobherrington**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This differs from https://github.com/thepracticaldev/dev.to/pull/6053 in that it only implements the vomitting buttons on the /mod view. Not on the user's profile.

One, to avoid dealing with the edge cached user/show view, and two, to avoid having this functionality live in two places.

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

image

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

giphy

### CODE REVIEW

File: app/controllers/reactions_controller.rb [hunk 1]
```diff
@@ -63,14 +63,12 @@ def create
...
       )
-      if reaction.save
-        result = "create"
-        Moderator::SinkArticles.call(reaction.user_id) if vomit_reaction_on_user?(reaction)
-        Notification.send_reaction_notification(reaction, reaction.reactable.user)
-        Notification.send_reaction_notification(reaction, reaction.reactable.organization) if organization_article?(reaction)
-      else
-        return render json: { error: reaction.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity
-      end
+      return render json: { error: reaction.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity unless reaction.save
```
Comment (citizen428) [inline app/controllers/reactions_controller.rb]: Personally I'm not a big fan of modifier `if`/`unless` on long lines, it's easy to overlook. This seems clearer to me:

```ruby
unless reaction.save
  render json: { error: reaction.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity
  return
end
```
Just a suggestion though, don't let it stop you 😃

Comment (citizen428) [reply app/controllers/reactions_controller.rb]: Also: Rails has a [`to_sentence`](http://api.rubyonrails.org/v5.2.4/classes/Array.html#method-i-to_sentence) method that comes in handy in cases like this:

```ruby
reaction.errors.full_messages.to_sentence
```

Comment (rhymes) [reply app/controllers/reactions_controller.rb]: TIL about `to_sentence`!

File: app/views/moderations/mod.html.erb [hunk 2]
```diff
@@ -24,18 +25,26 @@
...
+  </p>
+  <div class="tag-mod-form">
+    <h2>Flag <a href="<%= @moderatable.user.path %>"><%= @moderatable.user.username %></a> as Abusive</h2>
+    <button class="reaction-button <%= Reaction.cached_any_reactions_for?(@moderatable.user, current_user, "vomit") ? "reacted" : "" %>" data-reactable-id="<%= @moderatable.user.id %>" data-category="vomit" data-reactable-type="<%= @moderatable.user.class.name %>">
+      <%= image_tag("emoji/emoji-one-nausea-face-gray.png", alt: "Nausea face emoji") %>
+      <img class="reacted-emoji" src="<%= asset_path("emoji/emoji-one-nausea-face.png") %>" alt="Nausea down emoji" />
+    </button>
     <p>
-      Use <b>vomit</b> for code of conduct violations (harassment, being a jerk, spam, etc.).
+      This is the equivalent of vomitting on <b>all</b> of this user's articles.
```
Comment (rhymes) [inline app/views/moderations/mod.html.erb]: `vomiting` with one t?

File: app/controllers/reactions_controller.rb (hunk 1 above)
Comment (rhymes) [inline app/controllers/reactions_controller.rb]: this line is super long :(

File: app/views/moderations/mod.html.erb [hunk 3]
```diff
@@ -24,18 +25,26 @@
...
+  </p>
+  <p>
+    Use <b>thumbsdown</b> to move this content "down" for any reason (quality, usefulness, etc.).
+  </p>
+  <p>
+    Use <b>vomit</b> for code of conduct violations (harassment, being a jerk, spam, etc.).
+  </p>
+  <div class="tag-mod-form">
+    <h2>Flag <a href="<%= @moderatable.user.path %>"><%= @moderatable.user.username %></a> as Abusive</h2>
+    <button class="reaction-button <%= Reaction.cached_any_reactions_for?(@moderatable.user, current_user, "vomit") ? "reacted" : "" %>" data-reactable-id="<%= @moderatable.user.id %>" data-category="vomit" data-reactable-type="<%= @moderatable.user.class.name %>">
```
Comment (rhymes) [inline app/views/moderations/mod.html.erb]: this line is super long :D I suspect you use word wrapping on your editor, right? :D

File: app/controllers/reactions_controller.rb (hunk 1 above)
Comment (jacobherrington) [reply app/controllers/reactions_controller.rb]: Rubocop really wanted this to be a guard clause, but I'll see if I can make it readable.

File: app/controllers/reactions_controller.rb [hunk 4]
```diff
@@ -63,14 +63,16 @@ def create
...
-      else
-        return render json: { error: reaction.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity
+
+      unless reaction.save
+        render json: { error: reaction.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity
+        return
       end
+
+      result = "create"
+      Moderator::SinkArticles.call(reaction.user_id) if vomit_reaction_on_user?(reaction)
```
Comment (Zhao-Andy) [inline app/controllers/reactions_controller.rb]: This should be `reactable_id`:
```suggestion
      Moderator::SinkArticles.call(reaction.reactable_id) if vomit_reaction_on_user?(reaction)
```

File: app/controllers/reactions_controller.rb [hunk 5]
```diff
@@ -63,14 +63,16 @@ def create
...
-        result = "create"
-        Moderator::SinkArticles.call(reaction.user_id) if vomit_reaction_on_user?(reaction)
-        Notification.send_reaction_notification(reaction, reaction.reactable.user)
-        Notification.send_reaction_notification(reaction, reaction.reactable.organization) if organization_article?(reaction)
-      else
-        return render json: { error: reaction.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity
+
+      unless reaction.save
+        render json: { error: reaction.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity
+        return
```
Comment (Zhao-Andy) [inline app/controllers/reactions_controller.rb]: Nice refactor 👍

File: app/controllers/reactions_controller.rb (hunk 1 above)
Comment (citizen428) [reply app/controllers/reactions_controller.rb]: @jacobherrington The trick is to put two statements into the `unless`, then it shouldn't suggest to make it a guard 😉 Unless it got smarter over the years, I actually haven't tried that.

**Pull request merged**